### ISSUE TITLE
vfio: Fix INTx EOI with resample fd

### ISSUE THREAD

**Issue by sboeuf** [cloud-hypervisor/cloud-hypervisor#4999]
As suggested by @zsdaka [here](https://github.com/cloud-hypervisor/cloud-hypervisor/discussions/4995#discussioncomment-4385583), level triggered interrupts from INTx might not be correctly acknowledge, meaning they might never be unmasked from a VFIO standpoint.

The proper solution is to rely on KVM ability to register a resample eventfd when registering an irqfd.
With this approach, it would be completely transparent for the VMM to unmask the INTx for VFIO. Basically we would create and register a resamplefd, and then we would invoke `VFIO_IRQ_SET_DATA_EVENTFD | VFIO_IRQ_SET_ACTION_UNMASK`. This way, whenever an event is received on the `resamplefd`, this would automatically trigger the unmasking of the associated INTx.
That would require:
- the vfio-ioctls crate to be updated first as we must be able to register an eventfd associated with the unmasking action.
- the use of register_irqfd_with_resample from kvm-ioctls crate, by creating the associated function in the hypervisor crate from Cloud Hypervisor
- patching Cloud Hypervisor codebase to get rid of the BAR read/write deassertion mechanism and replace it with the resampling code

One thing to be careful about is that we should only use the resample fd when dealing with INTx, but not for MSI or MSI-X.

**Comment by sboeuf** [comment]
@zsdaka now that the first step is completed, are you going to take care of the second and third items?

**Comment by zsdaka** [comment]
> @zsdaka now that the first step is completed, are you going to take care of the second and third items?
OK. I'll go on the remaining items.

**Comment by likebreath** [comment]
> * [ ]  the use of register_irqfd_with_resample from kvm-ioctls crate, by creating the associated function in the hypervisor crate from Cloud Hypervisor

For the record, the proposed approach does not work because kvm does not allow to use `resamplefd` with userspace ioapic (e.g. split irqchip) [1]. This is why #5239 extended our (userspace) ioapic implementation to manually kick the `resamplefd` upon `KVM_EXIT_IOAPIC_EOI`.

[1] https://github.com/torvalds/linux/commit/654f1f13ea56b92bacade8ce2725aea0457f91c0

**Comment by rbradford** [comment]
Are issues with vfio and INTx EOI still relevant any more? Perhaps we can just go ahead and close this! @sboeuf WDYT?

**Comment by sboeuf** [comment]
> Are issues with vfio and INTx EOI still relevant any more? Perhaps we can just go ahead and close this! [@sboeuf](https://github.com/sboeuf) WDYT?

I agree this can be closed and can eventually be revisited if needed. But TBH, most PCI cards CH intends to support should have at least MSI supported.
